README.md
# OpenShift Release Images
This repository provides a subscription that will populate all the latest OpenShift images into Advanced Cluster Management for OpenShift deployments. More information about OpenShift release images and the channels mentioned below can be found here: https://docs.openshift.com/container-platform/4.4/updating/updating-cluster-between-minor.html#understanding-upgrade-channels_updating-cluster-between-minor

# Repository layout
- `fast` `./clusterImageSets/fast/*` : The fast channel OpenShift releases, generally available and fully supported
- `stable` `./clusterImageSets/stable/*` : The latest stable OpenShift releases, same as fast, but with connected customer feedback
- `candidate` `./clusterImageSets/releases/*` : All release images are in this folder, those labelled `candidate` releases are not supported

See the `Custom curated` section on controlling your own OpenShift release timelines with Advanced Cluster Management

## Latest supported images (ONLINE)
- Populates the latest OpenShift `fast` release images
- Run the following command
```bash
# Connect to you Red hat Advanced Cluster Management hub
# On OpenShift Dedicated, run "oc new-project hive-clusterimagesets"
make subscribe-fast
```
- After about 60s the Create Cluster console will list the latest supported OpenShift images
### Stable channel images
- Populates the latest 2x `stable` release images
```bash
make subscribe-stable
```

### Alternate full CLI command
- Populates the latest `fast` release images
```bash
oc apply -k subscribe/
```

### How to pause the subscription
#### Prerequisites
1. The fast channel subscription has been applied
2. Logged into the ACM hub
#### Pause the Fast Channel subscription
```bash
make pause-fast

# Full CLI command:
oc -n hive patch appsub hive-clusterimagesets-subscription-fast-0 --type='json' -p='[{"op":"replace","path": "/metadata/labels/subscription-pause","value":"true"}]'
```
#### Unpause the Fast Channel subscription
```bash
make unpause-fast

# Full CLI command:
oc -n hive patch appsub hive-clusterimagesets-subscription-fast-0 --type='json' -p='[{"op":"replace","path": "/metadata/labels/subscription-pause","value":"false"}]'
```
_Note: `hive-clusterimagesets-subscription-stable-0` resource name can be substituted, in the CLI pause commands, if you are working with `stable` release images._

### Continuous updates
- This repository periodically updates as new fast and stable release images are minted
- Changes in this repository will be applied to your subscribed cluster
- This is the stable channel list being followed by this repository, [link](https://github.com/openshift/cincinnati-graph-data/blob/master/channels/stable-4.3.yaml)

### Uninstall
```bash
make unsubscribe-all

# Full CLI commands:
oc delete -k subscribe/
oc delete -f subscribe/subscription-stable.yaml  #If your using the stable channel
```

## Custom curated (ONLINE)
- Fork this repository
- Update the the `./subscribe/channel.yaml` file, changing the organization `open-cluster-management` to your `organization_name` or `github_username` where you forked the repository.
```yaml
spec:
  type: GitHub
  pathname: https://github.com/NAME_or_ORGANIZATION/acm-hive-openshift-versions.git
```
- Place the YAML files for the images you want to appear in the Red Hat Advanced Cluster Management console under `./clusterImageSets/stable/*` or `./clusterImageSets/fast/*`
- Commit and push your changes to the forked repository
- Run the following command
```bash
make subscribe-fast   #fast channel
make subscribe-stable #stable channel
```
- After about 60s the Create Cluster console will list the new images available from your forked repository
- Add new OpenShift install images by created additional files in the `clusterImageSets/stable/` or `clusterImageSets/fast/` directories

## How to get new versions
- This repository will automatically update with the latest stable and fast versions
- You can monitor this repository and merge changes to your forked repository
- As soon as new images are committed to this repository and merged to your fork, they will become available in the Red Hat Advanced Cluster Management console (about 60s)
- This is the install image repository being used: https://quay.io/repository/openshift-release-dev/ocp-release?tab=tags
- These are the streams being followed by this repository: https://github.com/openshift/cincinnati-graph-data/blob/master/channels/

## (Technical Preview) Usecase - OFFLINE - limited images
- Copy the `clusterImageSets` directory to a system that has access to the disconnected Red Hat Advanced Cluster Management Hub
- Delete the YAML files for OpenShift versions you do not want to host OFFLINE
- Modify the `clusterImageSet` YAML files for the remaining OpenShift release images to point to the correct `OFFLINE` repository
```yaml
apiVersion: hive.openshift.io/v1
kind: ClusterImageSet
metadata:
    name: img4.4.0-rc.6-x86-64
spec:
    releaseImage: IMAGE_REGISTRY_IPADDRESS_or_DNSNAME/REPO_PATH/ocp-release:4.4.0-rc.6-x86_64
```
- Make sure the images are loaded in the OFFLINE image registry referenced in the YAML
- Apply a subset of the YAML files
```bash
oc apply -f clusterImageSets/CHANNEL/VERSION/FILE_NAME.yaml
```
- The Create Cluster console will list only the images available from the `cluseterImageSets` directory

## Secure Github repostiory
- Uncomment the secret reference in `subscribe/channel.yaml`
- Create a `subscribe/secret.yaml` file with the following contents
```yaml
  ---
  apiVersion: v1
  kind: Secret
  metadata:
    name: my-github-secret
    namespace: ocp-clusterimagesets
  data:
    user: BASE64_ENCODED_GITHUB_USERNAME
    accessToken: BASE64_ENCODED_GITHUB_TOKEN
```
- The following command is used to encode base64: `echo "VALUE_TO_ENCODE" | base64`  place the output in the yaml file.
- Create the secret
```bash
make subscribe-fast
# OR
make subscribe-stable
```

# Development, to support a new OpenShift version via Travis
Update the Travis variable: `LIST_VERSIONS`
Make sure to enclose the space separated version numbers with quotes.
```
LIST_VERSIONS = "4.3 4.4 4.5"
```

ChangeLog.md
# 2020-09-16
- Migrated ./subscription to ./subscribe
- Updated the subscription to be a fully compliant ACM2.1 Application (Used ACM Console to generate YAML)
- Updated the README.md for the directory change to ./subscribe
- Added `4.6` builds to Travis, once OCP promotes the first fast, it will be automatically published

# (OLD)
## ToDo's
1. <s>Add a Travis job that will keep the images up to date automatically</s>
2. <s>Add a time window to the Subscription</s>
3. Autoamted test:
   a. AWS
   b. GCP
   c. Azure
   d. Bare Metal


tooling/create-ocp-clusterimagesets.py
# Author: jnpacker

import requests
import os
import os.path

# To support new versions, add them to this array that is set for the Travis job LIST_VERSIONS="4.3 4.4 4.5"
VERSIONS = os.environ.get("LIST_VERSIONS").split(" ")
for version in VERSIONS:
    print(" Checking for release images: " + version + ".x")
    if not os.path.isdir("clusterImageSets/releases/" + version):
        newDir = "clusterImageSets/releases/" + version
        print(" Create directory: " + newDir)
        os.mkdir(newDir)
    
    resp = requests.get('https://quay.io/api/v1/repository/openshift-release-dev/ocp-release/image/')
    if resp.status_code != 200:
        # There was a problem
        raise ValueError('GET quay.io status code: %s\n%s' % (resp.status_code, resp.text))
    
    # Loop through all images found in quay.io
    for image in resp.json()['images']:
        if (image['tags'] != []):
            tag=image['tags'][0]  # There is only one tag for now in each image
            if ("x86_64" in tag and version+"." in tag):
                print('Checking tag: {}'.format(tag), end='')

                # Check if we already have the file in the stable, fast or releases channel
                fileName="img" + tag + ".yaml"
                fileNotFound=True
                #for channel in ["releases","stable","fast"]:
                #    if os.path.isfile("clusterImageSets/" + channel + "/" + version + "/" + fileName):
                #        fileNotFound=False
                #if fileNotFound:
                if not os.path.isfile("clusterImageSets/releases/" + version + "/" + fileName):
                    #imgName=tag.replace("x86_64","fast")
                    imgName=tag.replace("_","-")
                    yaml= open("clusterImageSets/releases/" + version + "/" + fileName,"w+")
                    cisr = ("---\n" +
                            "apiVersion: hive.openshift.io/v1\n"
                            "kind: ClusterImageSet\nmetadata:\n"
                            "  name: img" + imgName + "\n"
                            "  labels:\n"
                            "    channel: candidate\n"
                            "    visible: \"false\"\n"
                            "spec:\n"
                            "  releaseImage: quay.io/openshift-release-dev/ocp-release:" + tag + "\n")
                    yaml.write(cisr)
                    yaml.close()
                    print(" Created clusterImageSet")
                else:
                    print(" Skipped, already exists")
